### PR TITLE
Unify "draw per model" and "draw per instance"

### PR DESCRIPTION
**Pull request by gyagp**
This patch just merges the logic of "draw per model" and "draw per
instance", while keeping the previous code logic. We need further patches
to clean up code.

### CODE REVIEW

File: src/aquarium-optimized/Aquarium.cpp [hunk 1]
```diff
@@ -204,10 +205,8 @@ bool Aquarium::init(int argc, char **argv) {
   }
   std::bitset<static_cast<size_t>(TOGGLE::TOGGLEMAX)> availableToggleBitset =
       mContext->getAvailableToggleBitset();
-  if (availableToggleBitset.test(
-          static_cast<size_t>(TOGGLE::UPATEANDDRAWFOREACHMODEL))) {
-    toggleBitset.set(static_cast<size_t>(TOGGLE::UPATEANDDRAWFOREACHMODEL));
-  }
+  mDrawPerInstance =
+      availableToggleBitset.test(static_cast<size_t>(TOGGLE::DRAWPERINSTANCE));
```
Comment (hujiajie) [inline src/aquarium-optimized/Aquarium.cpp]: Can we keep `toggleBitset.set()` at the same time? No matter this has any actual effect or not, without it gives me the feeling that the backend and the frontend are out of sync.

File: src/aquarium-optimized/Aquarium.cpp [hunk 2]
```diff
@@ -786,42 +785,38 @@ void Aquarium::render() {
...
-    updateAndDrawFishes();
-    mContext->updateFPS(mFpsTimer, &mCurFishCount, &toggleBitset);
-  } else {
-    updateBackground();
-    updateFishes();
-    mContext->updateFPS(mFpsTimer, &mCurFishCount, &toggleBitset);
-
-    // Begin render pass
+  updateAndDrawBackground();
+  updateAndDrawFishes();
```
Comment (hujiajie) [inline src/aquarium-optimized/Aquarium.cpp]: When `mDrawPerInstance` is `false` they don't really draw, but the names sound overlapping with the following `draw{Background,Fishes}()`. Let's consider a less confusing proposal for them.

Comment (hujiajie) [reply src/aquarium-optimized/Aquarium.cpp]: Not sure if the following is possible:

```c++
void Aquarium::updateFishes(UpdateCallback cb) {
  // ...
}

void Aquarium::render() {
  // ...
  if (mDrawPerInstance) {
    updateFishes(drawFishTrampoline);
  } else {
    updateFishes(nullptr);
    drawFishes();
  }
  // ...
}
```

Comment (gyagp) [reply src/aquarium-optimized/Aquarium.cpp]: I will further refine this part in next patches.

File: src/aquarium-optimized/Aquarium.cpp (hunk 1 above)
Comment (gyagp) [reply src/aquarium-optimized/Aquarium.cpp]: Done

**Comment by gyagp** [comment]
I merged all the logic into a single function. @hujiajie, PTAL!

File: src/aquarium-optimized/Aquarium.cpp [hunk 3]
```diff
@@ -786,219 +782,112 @@ void Aquarium::render() {
...
-    }
-  }
+          ? MODELNAME::MODELBIGFISHBINSTANCEDDRAWS
+          : MODELNAME::MODELBIGFISHB;
 
-  mContext->updateAllFishData();
-}
+  for (int i = 0; i <= MODELNAME::MODELMAX; ++i) {
+    if (i >= MODELNAME::MODELSMALLFISHA && (i < fishBegin || i > fishEnd))
+      continue;
```
Comment (hujiajie) [inline src/aquarium-optimized/Aquarium.cpp]: Why

```c++
FOR_EACH_MEMBER(i) {
  if (IS_MEMBER_OF_X(i)) continue;
  if (IS_MEMBER_OF_Y(i)) { doSomething(i); }
  else /*IS_MEMBER_OF_Z(i)*/ { doSomethingElse(i); }
}
```

than

```c++
FOR_EACH_MEMBER_OF_Y(i) { doSomething(i); }
FOR_EACH_MEMBER_OF_Z(i) { doSomethingElse(i); }
```

File: src/aquarium-optimized/Aquarium.cpp [hunk 4]
```diff
@@ -786,219 +782,112 @@ void Aquarium::render() {
...
+        model->updateFishPerUniforms(
+            sin(xClock) * xRadius, sin(yClock) * yRadius + fishHeight,
+            cos(zClock) * zRadius, sin(xClock - 0.04f) * xRadius,
+            sin(yClock - 0.01f) * yRadius + fishHeight,
+            cos(zClock - 0.04f) * zRadius, scale,
+            fmod((g.mclock + ii * g_tailOffsetMult) * fishTailSpeed * speed,
+                 static_cast<float>(M_PI) * 2),
+            ii);
+
+        model->updatePerInstanceUniforms(worldUniforms);
```
Comment (hujiajie) [inline src/aquarium-optimized/Aquarium.cpp:866]: Should this also be guarded by `!drawPerModel`?

File: src/aquarium-optimized/Aquarium.h [hunk 5]
```diff
@@ -518,15 +516,8 @@ class Aquarium {
...
   void resetFpsTime();
-  void updateWorldMatrix(Model *model);
 
-  void updateAndDrawFishes();
-  void updateAndDrawBackground();
-
-  void updateBackground();
-  void drawBackground();
-  void updateFishes();
-  void drawFishes();
```
Comment (hujiajie) [inline src/aquarium-optimized/Aquarium.h:529]: Please also remove `updateWorldMatrixAndDraw()` and `updateWorldProjections()`.

File: src/aquarium-optimized/Aquarium.cpp (hunk 3 above)
Comment (gyagp) [reply src/aquarium-optimized/Aquarium.cpp]: I hoped to keep the same structure of for loop. But your proposal is more clean, so I changed the code.

File: src/aquarium-optimized/Aquarium.h (hunk 5 above)
Comment (gyagp) [reply src/aquarium-optimized/Aquarium.h:529]: done

File: src/aquarium-optimized/Aquarium.cpp (hunk 4 above)
Comment (gyagp) [reply src/aquarium-optimized/Aquarium.cpp:866]: done

**Pull request merged**